subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java

```diff
@@ -64,17 +64,48 @@ interface Identity {
      */
     WorkOutput execute(ExecutionRequest executionRequest);
 
+    /**
+     * Parameter object for {@link #execute(ExecutionRequest)}.
+     */
     interface ExecutionRequest {
+        /**
+         * The directory to produce outputs into.
+         * <p>
+         * Note: it's {@code null} for tasks as they don't currently have their own workspace.
+         */
         File getWorkspace();
 
+        /**
+         * For work capable of incremental execution this is the object to query per-property changes through;
+         * {@link Optional#empty()} for non-incremental-capable work.
+         * <p>
+         * Note that incremental-capable work can also be executed non-incrementally, but even then a
+         * {@link org.gradle.work.InputChanges} instance is avaialble, returning all inputs as changed.
+         */
         Optional<InputChangesInternal> getInputChanges();
 
+        /**
+         * Output snapshots indexed by property fron the previous execution;
+         * {@link Optional#empty()} is information about a previous execution is not available.
+         */
         Optional<ImmutableSortedMap<String, FileSystemSnapshot>> getPreviouslyProducedOutputs();
     }
 
+    /**
+     * The result of executing the user code.
+     */
     interface WorkOutput {
+        /**
+         * Whether any significant amount of work has happened while executing the user code.
+         * <p>
+         * What amounts to "significant work" is up to the type of work implementation.
+         */
         WorkResult getDidWork();
 
+        /**
+         * Implementation-specific output of executing the user code.
```

Comment:
Please document here what this being `null` means.

Reply:
It really depends on the implementation, like `Callable.call()` can return `null`, but the meaning depends on the implementation. So not sure what you want to add here.

Again, if we decide to do https://github.com/gradle/gradle/pull/22066, it will add a bit of explanation by turning this from `Object` to `T`.

Reply:
I suppose `null` means that there isn't an implementation specific output of executing the user code. If you think that adding a phrase like that doesn't add any extra value, then we can also leave it as is.